### PR TITLE
Change noobaa system reconcile to use "smart" desired state

### PR DESCRIPTION
**Pull request by nb-ohad**
- Restructure the code to use CreateOrUpodate (from controllerutils) which get, create and update, instead of doing it manually.

- Reconcile specific props of the NooBaa CR, stated by the setNooBaaDesiredState func, instead of blindly override everything in the NooBaa CR. (leaving some form of autonomy over the NooBaa CR).

### CODE REVIEW

File: pkg/controller/storagecluster/noobaa_system_reconciler.go [hunk 1]
```diff
@@ -22,16 +22,6 @@ import (
...
-	cephClusterCreated := false
-
-	err := controllerutil.SetControllerReference(sc, nb, r.scheme)
-	if err != nil {
-		return nil
-	}
-
 	// find cephCluster
 	foundCeph := &cephv1.CephCluster{}
 	err = r.client.Get(context.TODO(), types.NamespacedName{Name: generateNameForCephCluster(sc), Namespace: sc.Namespace}, foundCeph)
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/noobaa_system_reconciler.go]: Undefined Variable : `err`

File: pkg/controller/storagecluster/noobaa_system_reconciler.go [hunk 2]
```diff
@@ -42,63 +32,38 @@ func (r *ReconcileStorageCluster) ensureNoobaaSystem(sc *ocsv1.StorageCluster, r
...
+		reqLogger.Info("Waiting on ceph cluster to initialize before starting noobaa")
+		return nil
 	}
 
-	// check if this noobaa instance aleady exists
-	found := &nbv1.NooBaa{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: nb.ObjectMeta.Name, Namespace: sc.Namespace}, found)
+	// Take ownership over the noobaa object
+	nb := &nbv1.NooBaa{}
+	err := controllerutil.SetControllerReference(sc, nb, r.scheme)
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/noobaa_system_reconciler.go]: `err` should be defined above.

```suggestion
	err = controllerutil.SetControllerReference(sc, nb, r.scheme)
```

File: pkg/controller/storagecluster/noobaa_system_reconciler.go [hunk 3]
```diff
@@ -42,63 +32,38 @@ func (r *ReconcileStorageCluster) ensureNoobaaSystem(sc *ocsv1.StorageCluster, r
...
-		if !reflect.DeepEqual(nb.Spec, found.Spec) {
-			reqLogger.Info("Updating spec for NooBaa")
-			found.Spec = nb.Spec
-			err := r.client.Update(context.TODO(), found)
-			if err != nil {
-				reqLogger.Error(err, "Failed to update NooBaa system")
-				return err
-			}
-		}
+		return nil
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/noobaa_system_reconciler.go]: Why not 
```suggestion
		return err
```
Adding a comment would be helpful is this is desired.

File: pkg/controller/storagecluster/noobaa_system_reconciler.go [hunk 4]
```diff
@@ -42,63 +32,38 @@ func (r *ReconcileStorageCluster) ensureNoobaaSystem(sc *ocsv1.StorageCluster, r
...
+	}
 
-		objectRef, err := reference.GetReference(r.scheme, found)
-		if err != nil {
-			return err
-		}
-		objectreferencesv1.SetObjectReference(&sc.Status.RelatedObjects, *objectRef)
+	// Reconcile the noobaa state, creating or updating if needed
+	_, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, nb.(runtime.Object), func() error {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/noobaa_system_reconciler.go]: No new variable declared on left.

```suggestion
	_, err = controllerutil.CreateOrUpdate(context.TODO(), r.client, nb.(runtime.Object), func() error {
```

File: pkg/controller/storagecluster/noobaa_system_reconciler.go (hunk 1 above)
Comment (nb-ohad) [reply pkg/controller/storagecluster/noobaa_system_reconciler.go]: fixed

File: pkg/controller/storagecluster/noobaa_system_reconciler.go (hunk 2 above)
Comment (nb-ohad) [reply pkg/controller/storagecluster/noobaa_system_reconciler.go]: fixed

File: pkg/controller/storagecluster/noobaa_system_reconciler.go (hunk 4 above)
Comment (nb-ohad) [reply pkg/controller/storagecluster/noobaa_system_reconciler.go]: fixed

File: pkg/controller/storagecluster/noobaa_system_reconciler.go (hunk 3 above)
Comment (nb-ohad) [reply pkg/controller/storagecluster/noobaa_system_reconciler.go]: fixed

**Pull request merged**